test/function/entry.cpp

```diff
@@ -2395,6 +2395,164 @@ namespace {
     stumpless_free_all(  );
   }
 
+
+
+
+
+
+
+
+  // TEST( SetMessageWideStrTest, AsciiMessage ) {
+  //   struct stumpless_entry *entry;
+  //   const char *ascii_message;
+  //   const struct stumpless_entry *result;
+  //   const char *new_message;
+
+  //   entry = create_empty_entry(  );
+  //   ASSERT_NOT_NULL( entry );
+
+  //   ascii_message = load_corpus( "cstring/ascii" );
```

Comment:
What do I do to make them UTF-16, in the program writing` L"test-message"` works to make it `wide str`. I tried saving the UTF-8 files (same content) as UTF-16 but the program does not recognize it as wide string.

Reply:
The `iconv` utility is useful for this, assuming you have access to it. If not I believe it is possible to do this with Powershell, assuming that you're on a Windows box if not a Linux one.

However, now that I'm thinking about this more you will run into trouble with `wchar_t` not being portable to load from a file, due to potentially different endianness and type sizes between platforms. Creating a portable way to do this is beyond the scope of this change - for now you can simply copy the text out of the load_corpus files and use the `L"string"` format to make a literal in the tests instead.

My apologies for the switch on this.

Reply:
```
error: incompatible pointer types assigning to 'const wchar_t *' from 'const char *'
    new_message = stumpless_get_entry_message( entry );
                 
```
```
TEST( SetMessageWideStrTest, Utf16Message ) {
    struct stumpless_entry *entry;
    const wchar_t *utf16_message;
    const struct stumpless_entry *result;
    const wchar_t *new_message; // if I make this char then all sorts of error pops up

    entry = create_empty_entry(  );
    ASSERT_NOT_NULL( entry );

    utf16_message = L"没有错误\00";
    ASSERT_NOT_NULL( utf16_message );

    result = stumpless_set_entry_message_str_w( entry, L"没有错误\00" );
    EXPECT_EQ( entry, result );
    EXPECT_NO_ERROR;

    new_message = stumpless_get_entry_message( entry );
    EXPECT_NOT_NULL( new_message );
    EXPECT_NO_ERROR;
    EXPECT_STREQ( utf16_message, new_message );

    delete[] utf16_message;
    free( ( void * ) new_message );
    stumpless_destroy_entry_and_contents( entry );
    stumpless_free_all(  );
  }
```
I reckon there has to be a` get_entry_message_w`for this to work? or is there a way to change something in the function to make it work?

Reply:
You're absolutely right, and this is another snag I had not considered.

You can solve this in two ways, with the first being much more work: implement the function `stumpless_get_entry_message_w` that will return the message in `wchar` format (and of course add tests for this as well).

The second is to simply convert the string you used to set the message to UTF-8, either as a new literal without the `L` prefix, or using a dynamic conversion function. `stumpless_get_entry_message` will return the function in UTF-8 format, so you just need to convert one of the two to do a comparison.

Reply:
I do not understand the second one, if the message I give is in UTF-8 then `stumpless_set_entry_message_str_w` would not accept it right?

Reply:
Correct, but the result returned by the get message function will be in UTF-8. So you will need to either convert it or the expected string to also be in UTF-8

Reply:
So after the `result` variable is computed I manually declare entry as "没有错误\00" would that do the trick?

Reply:
You won't need to manually add the null character, but yes that is the string literal that you could use to compare with the result of get message after the `_w` call is done, assuming you used the same characters but in wide form.

Reply:
Its not working out, I get gtest errors and various other errors.

Reply:
```
 result = stumpless_set_entry_message_str_w( entry, L"没有错误" );
    EXPECT_EQ( entry, result );
    EXPECT_NO_ERROR;

    entry = "没有错误";
    new_message = stumpless_get_entry_message( entry );
    EXPECT_NOT_NULL( new_message );
    EXPECT_NO_ERROR;
    EXPECT_STREQ( utf16_message, new_message );
```
Just to make sure I understood it right.

Reply:
If the first method is the only way to make things work, could you guide me on what changes must be made? If you could describe it as you do when you write the issues, would be very helpful.

Reply:
Your above code snippet looks mostly right, though you've overwritten the `entry` variable with a string literal before calling `stumpless_get_entry_message`. That will probably cause some errors, though I'm not sure if they're the ones you are encountering. If you still run into problems after this, please provide the errors and your current code and I'll take a look at the specific fault.

Reply:
I thought overwriting the entry variable like I did is what you meant. If that's not it then could you explain how else should entry variable be written?

Reply:
```
 result = stumpless_set_entry_message_str_w( entry, L"没有错误" );
    EXPECT_EQ( entry, result );
    EXPECT_NO_ERROR;

    // what should come here?
    new_message = stumpless_get_entry_message( entry );
    EXPECT_NOT_NULL( new_message );
    EXPECT_NO_ERROR;
    EXPECT_STREQ( utf16_message, new_message );
```

Reply:
Try something like this:

```c
const wchar_t *wide_message= L"没有错误";
const char *utf8_message = "没有错误";
const char *new_message;

result = stumpless_set_entry_message_str_w( entry, wide_message );
EXPECT_EQ( entry, result );
EXPECT_NO_ERROR;

new_message = stumpless_get_entry_message( entry );
EXPECT_NOT_NULL( new_message );
EXPECT_NO_ERROR;
EXPECT_STREQ( utf8_message, new_message );
```

Reply:
```
Expected equality of these values:
  entry
    Which is: 0x562541653640
  result
    Which is: NULL
/home/rmknan/Github/Stump2/stumpless/test/function/entry.cpp:2530: Failure
Value of: stumpless_has_error( )
  Actual: true
Expected: false
an unexpected error occurred: STUMPLESS_INVALID_ENCODING: could not convert the provided wide char string to a multibyte string (errno after the failed call: 84)
/home/rmknan/Github/Stump2/stumpless/test/function/entry.cpp:2535: Failure
Expected equality of these values:
  utf8_message
    Which is: "\xE6\xB2\xA1\xE6\x9C\x89\xE9\x94\x99\xE8\xAF\xAF"
    As Text: "没有错误"
  new_message
    Which is: "fixture message"
<end of output>
```
```
TEST( SetMessageWideStrTest, Utf16Message ) {
    struct stumpless_entry *entry;
    const wchar_t *utf16_message;
    const struct stumpless_entry *result;
    const char *utf8_message = "没有错误";
    const char *new_message;

    entry = create_empty_entry(  );
    ASSERT_NOT_NULL( entry );

    utf16_message = L"没有错误";
    ASSERT_NOT_NULL( utf16_message );
    // ASSERT_NOT_NULL( utf8_message );

    result = stumpless_set_entry_message_str_w( entry, utf16_message );
    EXPECT_EQ( entry, result );
    EXPECT_NO_ERROR;

    new_message = stumpless_get_entry_message( entry );
    EXPECT_NOT_NULL( new_message );
    EXPECT_NO_ERROR;
    EXPECT_STREQ( utf8_message, new_message );

    delete[] utf16_message;
    free( ( void * ) new_message );
    stumpless_destroy_entry_and_contents( entry );
    stumpless_free_all(  );
  }
```

Reply:
Based on the error, it looks like the test is working properly! Now you'll need to find out why that encoding error is happening, which is likely an issue with the implementation instead of the test.

Keep in mind that the issue may be in the existing wstring code, so be sure to consider that code as well as you work through the issue.

Reply:
Is there a way to debug the code like line by line? I mean if I could place a marker on the wstring code it could lead me to the individual piece of code which is failing. 

I see that the wstring code calls for three different h files. 

Could you also tell me what does these two stand for` HAVE_WINDOWS_H` and` HAVE_WCSRTOMBS_S`?

Reply:
A standard debugger will let you step through your code line by line, `gdb` being one of the most prominent choices. Many IDEs come with debugger integrations that make this easier to navigate, but that depends on your IDE.

The `HAVE_` symbols describe whether given functions and/or headers are available in the build system. Those two indicate whether `<windows.h>` and the function `wcsrtombs_s` are present, respectively.

Reply:
Out of the 3 header files being called, the wcsrtomb is the one being used by the program, and its definition is 
```

char *
no_wcsrtombs_s_copy_wstring_to_cstring( const wchar_t *str, int *copy_size );
```
I tried calling the other header files instead of this one but they gave out various errors as well.

Currently going through the `stumpless/src/config/no_wcsrtombs_s.c` file.

Reply:
Hmm, I'm not sure what the reason for the segfault is - if you just run it with `gdb` or a similar debugger you should be able to get a backtrace to where the failure occurred. This would be my normal next step.

If you continue to run into roadblocks and don't know where to go next, push what you have and let me know, in which case I'll work out the issues and co-author this change with you. There is also no particular rush on this right now, so if you'd like sole authorship you can keep working through it.

Reply:
Could you tell me how exactly I debug using gdb? I tried to compile the `stumpless/src/config/no_wcsrtombs_s.c` but it gives me an error saying cannot find the` include files`.

Reply:
A detailed tutorial on debugging is beyond the scope of what I can provide. At a high level, make sure that you configure in debug mode (`-DCMAKE_BUILD_TYPE=Debug` provided to the cmake), and then run `gdb` from the source directory. Here is some output from an example debug session, demonstrating how to set a breakpoint in your test, step into library code, and print information at one point in the source.

```
$ gdb ~/build-debug/function-test-entry
GNU gdb (Ubuntu 8.1.1-0ubuntu1) 8.1.1
Copyright (C) 2018 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from ~/build-debug/function-test-entry...done.
(gdb) b test/function/entry.cpp:97
Breakpoint 1 at 0x2a254: file test/function/entry.cpp, line 97.
(gdb) run
Starting program: ~/build-debug/function-test-entry
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Running main() from ~/build-debug/gtest/src/gtest/googletest/src/gtest_main.cc
[==========] Running 172 tests from 44 test suites.
[----------] Global test environment set-up.
[----------] 92 tests from EntryTest
[ RUN      ] EntryTest.AddElement

Breakpoint 1, (anonymous namespace)::EntryTest_AddElement_Test::TestBody (this=0x84dbd20) at test/function/entry.cpp:97
97          element = stumpless_new_element( "test-new-element" );
(gdb) step
stumpless_new_element (name=0x8083f7a "test-new-element") at src/element.c:478
478     stumpless_new_element( const char *name ) {
(gdb) next
483       VALIDATE_ARG_NOT_NULL( name );
(gdb)
485       if( unlikely( !validate_element_name( name, &name_length ) ) ) {
(gdb)
489       element = alloc_mem( sizeof( *element ) );
(gdb)
490       if( unlikely( !element ) ) {
(gdb) p element
$1 = (struct stumpless_element *) 0x84de0b0
(gdb)
```

Reply:
Based on the debugging of the code, the following code fails and the fail condition is met. (`no_wcsrtombs_s.c`)

```
conversion_result = wcsrtombs( NULL, &str, 0, &state );
  if( conversion_result == -1 ) {
    raise_wide_conversion_failure( errno, L10N_ERRNO_ERROR_CODE_TYPE );
    goto fail;
  }
```
[Stump Debug Sess.zip](https://github.com/goatshriek/stumpless/files/12822959/Stump.Debug.Sess.zip)

I have attached what the debug session goes through if you would like to have a look. Its not able to find several files that are associated with  `wcsrtombs` and therefore is not able to do the conversion from my understanding.

Reply:
Thank you for providing a debug session to review. The errors regarding missing files are simply reflections of the debugger not having access to standard library files - they are expected if you try to step into a library function like `wcsrtombs`. When you encounter a source line with such a call, often you want to use `next` instead of `step` to step over it.

I was able to reproduce the error you are encountering locally based on the current state of this pull request. After some quick inspection, I was able to resolve the error by initializing the `mbstate_t` conversion state, a step that was left out and is causing problems. After adding this, the current tests pass. The new function looks like:

```c
char *
no_wcsrtombs_s_copy_wstring_to_cstring( const wchar_t *str, int *copy_size ) {
  size_t buffer_size;
  char *buffer;
  mbstate_t state;
  size_t conversion_result;

  memset( &state, 0, sizeof( state ) ); // this is the new line (you'll also need to add an include of string.h)
  conversion_result = wcsrtombs( NULL, &str, 0, &state );
  if( conversion_result == -1 ) {
    raise_wide_conversion_failure( errno, L10N_ERRNO_ERROR_CODE_TYPE );
    goto fail;
  }
```

Give this a shot and see if it works for you.

Reply:
Yesterday's commit had the `UTF16 check` commented out. There was some confusion and It seems I had reset the changes to previous commit.

Apologies for that. I have pushed the new commit with the changes now. I tried adding the above piece of code,The build check still says `segfault` and the wide conversion is not happening.

Reply:
Thank you for the update. The segfault is happening because of the use of  `delete` and `free` on string literals - remove both of those calls from the end of the UTF16 test.

You will also likely need to call `setlocale(LC_ALL, "");` at the beginning of your test so that the conversion does not fail due to a locale issue. This should not be done in the library code itself as it should not modify locale settings, but must be done in order for the test conversion to succeed.

Finally, I would rename the test so that it refers to wide characters instead of UTF16. While debugging I noticed that my system uses UTF32 for wide characters, so using a name referring to UTF16 would be incorrect in some cases.

Reply:
> You will also likely need to call `setlocale(LC_ALL, "");`

So Where would i call it?

Reply:
Call it at the beginning of your `TEST` function in `entry.cpp`. You will also need to add the appropriate include, of course.

Reply:
```
home/rmknan/Github/Stump2/stumpless/test/function/entry.cpp:2535: Failure
Value of: (new_message) == __null
  Actual: true
Expected: false
/home/rmknan/Github/Stump2/stumpless/test/function/entry.cpp:2536: Failure
Value of: stumpless_has_error( )
  Actual: true
Expected: false
an unexpected error occurred: STUMPLESS_MEMORY_ALLOCATION_FAILURE: a memory allocation call failed
/home/rmknan/Github/Stump2/stumpless/test/function/entry.cpp:2537: Failure
Expected equality of these values:
  utf8_message
    Which is: "\xE6\xB2\xA1\xE6\x9C\x89\xE9\x94\x99\xE8\xAF\xAF"
    As Text: "没有错误"
  new_message
    Which is: NULL
```

The `Segfault` is solved but the conversion  still fails.

Reply:
If you push the current state I will take a look. Note that you also need to merge in the `latest` branch to your local branch, as some conflicts have arisen since you started.

Reply:
Were you able to take a look?

Reply:
You've left a large portion of `no_wcsrtombs_s_copy_wstring_to_cstring` commented out, which is part of the problem.

You should be seeing compilation errors regarding passing a size_t pointer into the wstring to cstring function, which takes an int pointer. Correct this error (by changing `new_message_length` to an int) and the problem should resolve (at least it did for me).

Reply:
I was trying something by commenting those out but did not work. The errors have been resolved, did not notice the int `new_message_length`. Thank you, I corrected the other two test the ASCII and Long ASCII as well and it runs successfully.